knox/views.py

```diff
@@ -79,3 +79,33 @@ def post(self, request, format=None):
         user_logged_out.send(sender=request.user.__class__,
                              request=request, user=request.user)
         return Response(None, status=status.HTTP_204_NO_CONTENT)
+
+
+class TokenRefreshView(APIView):
+
+    authentication_classes = (TokenAuthentication,)
+    permission_classes = (IsAuthenticated,)
+
+    def post(self, request, format=None):
+        return Response(
```

Comment:
We should do an explicit update on the expiry here:

- If we don't, in the context of my previous comments, it will not work for people with `AUTO_REFRESH=False` and wanting to manually refresh.

- this view should bypass `MIN_REFRESH_INTERVAL` as it is an explicit request for a refresh and not an implicit one triggered by `AUTO_REFRESH`.

Ideally, we should find a way to disable the behavior of `AUTO_REFRESH` in the authentication class when hitting this endpoint.

Reply:
@johnraz I agree with you, ill go back to the drawing board a bit to try to make this more explicity and flexible.

Reply:
@johnraz I am not sure how to circumvent the auto refresh in auth without making it messy. I mean we could explicitly do it in the refresh view if `AUTO_REFRESH` is not set, but if we do that it will not bypass `MIN_REFRESH_INTERVAL` if the user happen to be using `AUTO_REFRESH` also.. which is kinda unexpected behaivor if you ask me..

Reply:
So the auto-refresh flow today is:
- hit a view with a token -->  if `MIN_REFRESH_INTERVAL` allows it, update expiry and hit the DB with a write.
- hit a view with a token -->  if `MIN_REFRESH_INTERVAL` denies it, don't update expiry and don't write to the DB.

What we'd ideally want for the refresh view (whether or not auto-refresh is active) is:
- hit the refresh view with a token --> refresh the token in every case by doing a single write to the DB

If auto-refresh is off, there is no issue.

If auto-refresh is on, hitting the refresh view will do this:
- if `MIN_REFRESH_INTERVAL` is expired, the auto-refresh will update the expiry and write to the DB at authentication time, next the view will do it again, resulting in 2 writes for nothing.
- if `MIN_REFRESH_INTERVAL` is not expired, only the view will write to the DB, 1 write everybody is happy.

The problem we need to solve is to avoid to have 2 consecutive useless writes to the DB...
Sadly I don't think we have access to the view or anything useful to identify the refresh view in the authenticate method (where `AUTO_REFRESH` happens)...

I would say, let the 2 DB write be and add a note in the doc that if you use `AUTO_REFRESH` and the refresh view together, a double write could happen.

Reply:
Taking the above statement back.
The view can be accessed in the authenticate method by doing:
```
request.parser_context['view']
```

This means we could simply add a property to the class like:
```
class RefreshView():
    disable_auto_refresh=True
```

And then check for that in the `authenticate` / `authenticate_credentials` method of the `TokenAuthentication` to disable auto-refresh "per view"

Reply:
@johnraz, wonderful research, thank you. I have been experimenting with this approach, but it seems to be not as trivial as it initially seemed. 

If we reference `AUTO_REFRESH` value in the authenticate/_credentials() methods it needs to be accessible on every view which kinda sucks because then we need to add it to each view or do some weird complex thing of trying to validate who or which part of the code is calling us -- which I dont think we should do since this is a library. 

I dont like that so maybe we should make a `KnoxAPIView` class that extends from `APIView` with the only difference this attribute `AUTO_REFRESH` is added as a field attribute like `auto_refresh` or so.

But I am not sure if its a good approach, though I would like to be able to do away with things like `if knox_settings.TOKEN_TTL` and only have `if token_ttl` or `auto_refresh` directly as an attribute in the class. Maybe im overthinking, but thats the current status. I will ponder this some more.

Reply:
> If we reference AUTO_REFRESH value in the authenticate/_credentials() methods it needs to be accessible on every view

I think we could solve this with something like:
```
if getattr(self, AUTO_REFRESH, knox_settings.AUTO_REFRESH):
    self.renew_token(auth_token)
```

instead of the current:
https://github.com/James1345/django-rest-knox/blob/c1972fcddf86179680522d6af1b463b02c49b00c/knox/auth.py#L77-L78

That way the order of precedence would be:
Class attribute's value first and next settings value, which is what we want.

Reply:
Thats a very nice solution, I like it. This PR does add some complexity to testing, I think I will have to mock some parts which tests the TokenAuthentication class directly because in that context there wont be any view. Is it okay if I mock it? no tests currently employs mocking.

Reply:
Cool! Glad you made it work 👍